web/source/kmwembedded.ts

```diff
@@ -361,14 +362,27 @@ namespace com.keyman.text {
       /* Clear any pending (non-popup) key */
       osk.vkbd.keyPending = null;
 
-      // Changes for Build 353 to resolve KMEI popup key issues      
+      // Changes for Build 353 to resolve KMEI popup key issues
       keyName=keyName.replace('popup-',''); //remove popup prefix if present (unlikely)
 
+      let displayLayer = keymanweb.core.keyboardProcessor.layerId;
+      let functionalLayer = '';
+
+      // Regex for 'display layer'-'virtual key name'+'optional functional layer'
       // Can't just split on '-' because some layers like ctrl-shift contain it.
-      let separatorIndex = keyName.lastIndexOf('-');
+      // Virtual key can be T_, U_, K_, or ISO 9995
+      let rx = new RegExp("(.*)-([TKU]_[^+]*|[A-E]\\d\\d)\\+?(.*)?");
+      let matches = keyName.match(rx);
+      displayLayer = matches[1];
+      keyName = matches[2];
+      if (matches[3]) {
+        // Optional function layer
+        functionalLayer = matches[3]
+        keyName += '+' + functionalLayer;
+      }
 
-      if (separatorIndex > 0) {
-        keyName = keyName.substring(separatorIndex+1);
+      if(displayLayer == 'undefined') {
+        displayLayer=keymanweb.core.keyboardProcessor.layerId;
```

Comment:
What is display layer used for once it has been set?

Reply:
It had more use in 14.0 code (#5651) where it was used to determine `layer` and then get the modifier state
```
// Determine the "functional" layer
      let layer = displayLayer;
      separatorIndex = keyName.lastIndexOf('+');
      if (separatorIndex > 0) {
        layer = keyName.substring(separatorIndex+1);
      }

      // Note:  this assumes Lelem is properly attached and has an element interface.
      // Currently true in the Android and iOS apps.
      var Lelem=keymanweb.domManager.getLastActiveElement(),keyShiftState=com.keyman.text.KeyboardProcessor.getModifierState(layer);
```

Lastly in 14.0, it was used to:
```
      // Check the virtual key 
        Lcode: Codes.keyCodes[keyName],
        LisVirtualKey: true,
        kName: keyName,
        kLayer: layer,
        kbdLayer: displayLayer,
        ...
```